lua/orgmode/treesitter/headline.lua

```diff
@@ -205,12 +205,15 @@ function Headline:set_todo(keyword)
   local current_todo = self:todo()
   if current_todo then
     tree_utils.set_node_text(current_todo, keyword)
+    self:update_cookie()
     return
   end
 
   local stars = self:stars()
   local text = ts.get_node_text(stars, 0)
   tree_utils.set_node_text(stars, string.format('%s %s', text, keyword))
+  self:refresh()
+  self:update_cookie()
```

Comment:
Why is this necessary? And if we `update_cookie()` here, is there even a need to call it in L208?

Reply:
The L208 call is necessary because of the `return` statement at L209. The code was written that way when I added the calls to `update_cookie()`. Prior to my PR, I think the function worked as follows: if the headline is a todo item already, set the node text to `keyword` and then return; if not, insert the TODO keyword into the node. `self:update_cookie()` needs to be called after both changes, and there's probably a way to refactor the function to only call it once, but I didn't feel qualified to rewrite the original logic of the function in this PR.

As to the other question about `self.refresh()`: the short answer is "because it doesn't work without it." I don't 100% know why the call to `self:refresh()` is necessary at L215, but if it isn't there, the value of the headline is wrong when `self:update_cookie()` is called. I suspect it's because `tree_utils.set_node_text()` uses `vim.api.nvim_buf_set_text()` to update the node and something in `self:refresh()` causes treesitter to get back in sync so that the calls to the treesitter API in `self:update_cookie()` return the correctly updated value. It took my over an hour of debugging to settle on `self:refresh()` here and I tried several other solutions (including `vim.schedule_wrap()` and `vim.defer_fn()`), so I'm pretty confident it has to be called here (unless there's some other solution I didn't try). But I agree with the original question: that call to `self:refresh()` looks wrong.

Reply:
> because of the return statement at L209

:man_facepalming: of course, my bad

> because it doesn't work without it

Let's figure out what is going on and then document the code